opencog/reasoning/RuleEngine/rule-engine-src/pln/FCMemory.cc

```diff
@@ -85,9 +85,23 @@ Handle FCMemory::get_cur_target(void) {
 
 bool FCMemory::isin_target_list(Handle h) {
 	for (Handle hi : target_list_) {
-		//TODO recursive search might be the right way
 		if (hi.value() == h.value())
 			return true;
+		//recursive lookup
+		else if (LinkCast(hi)) {
+			HandleSeqSeq hseqs;
+			hseqs.push_back(as_->getOutgoing(hi));
+			do {
+				HandleSeq iset = hseqs[hseqs.size() - 1];
+				hseqs.pop_back();
+				for (Handle i : iset) {
+					if (i.value() == h.value())
+						return true;
+					else if (LinkCast(i))
+						hseqs.push_back(as_->getOutgoing(i));
+				}
+			} while (not hseqs.empty());
+		}
 	}
```

Comment:
FYI, there are a variety of  simple minor utilities  in PatternUtils.h that do similar things -- finding atoms in trees of other atoms, etc.